ts/sagas/startup/loadVisibleServicesHandler.ts

```diff
@@ -35,17 +42,83 @@ export function* loadVisibleServicesRequestHandler(
     );
 
     if (response !== undefined && response.status === 200) {
-      const { items } = response.value;
-      yield put(loadVisibleServices.success(items));
-      // Parallel fetch of those services that we haven't loaded yet
-      const servicesById: GlobalState["entities"]["services"]["byId"] = yield select<
-        GlobalState
-      >(_ => _.entities.services.byId);
-      const serviceIds = items.map(_ => _.service_id).filter(isDefined);
-      const newServiceIds = serviceIds.filter(
-        id => !isServiceLoaded(servicesById[id])
+      const { items: visibleServices } = response.value;
+      yield put(loadVisibleServices.success(visibleServices));
+
+      const visibleServiceVersionById = visibleServices.reduce<
+        VisibleServiceVersionById
+      >(
+        (accumulator, currentValue) => ({
+          ...accumulator,
+          [currentValue.service_id]: currentValue.version
+        }),
+        {}
+      );
+
+      const storedServicesById: ServicesByIdState = yield select(
+        servicesByIdSelector
+      );
+
+      const messagesIdsByServiceId: MessagesIdsByServiceId = yield select(
+        messagesIdsByServiceIdSelector
+      );
+
+      // Create an array of tuples containing:
+      // - serviceId (to remove service from the servicesById section of the redux store)
+      // - organizationFiscalCode (to remove service from serviceIdsByOrganizationFiscalCode
+      //   section of the redux store)
+      const serviceTuplesToRemove = Object.keys(storedServicesById).reduce<
+        ReadonlyArray<ITuple2<string, string | undefined>>
+      >((accumulator, serviceId) => {
+        // Check if this service id must be removed
+        // A service must be removed if is no more visible and not used by any loaded message.
+        const mustRemoveServiceId =
+          visibleServiceVersionById[serviceId] === undefined &&
+          messagesIdsByServiceId[serviceId] === undefined;
+
+        if (!mustRemoveServiceId) {
+          return accumulator;
+        }
+
+        const storedPotService = storedServicesById[serviceId];
+
+        if (storedPotService !== undefined) {
+          // If the service detail is also loaded get the organization fiscal code
+          const organizationFiscalCode = pot.getOrElse(
+            pot.map(storedPotService, _ => _.organization_fiscal_code),
+            undefined
+          );
+
+          return accumulator.concat(Tuple2(serviceId, organizationFiscalCode));
+        }
+
+        return accumulator;
+      }, []);
+      // Dispatch action to remove the services from the redux store
+      yield put(removeServiceTuples(serviceTuplesToRemove));
+
+      // Load only services that are not already stored or are outdated (we have a new version).
+      const serviceIdsToLoad = Object.keys(visibleServiceVersionById).filter(
```

Comment:
instead of extracting the keys from the map that you constructed from `visibleServices`, you should probably just filter `visibleServices`